app-backend/db/src/main/scala/TaskDao.scala

```diff
@@ -0,0 +1,208 @@
+
+  val insertF: Fragment =
+    fr"INSERT INTO " ++ tableF ++ fr"""(
+          id,
+          created_at,
+          created_by,
+          modified_at,
+          modified_by,
+          project_id,
+          project_layer_id,
+          status,
+          locked_by,
+          locked_on,
+          geometry
+     )
+     """
+
+  def updateF(taskId: UUID, update: Task.TaskFeatureCreate): Fragment =
+    fr"UPDATE " ++ tableF ++ fr"""SET
+      project_id = ${update.properties.projectId},
+      project_layer_id = ${update.properties.projectLayerId},
+      status = ${update.properties.status},
+      geometry = ${update.geometry}
+    WHERE
+      id = $taskId
+    """;
+
+  def appendAction(
+      taskId: UUID,
+      initialStatus: TaskStatus,
+      newStatus: TaskStatus,
+      userId: String
+  ): ConnectionIO[Int] =
+    if (initialStatus != newStatus) {
+      fr"""INSERT INTO task_actions (task_id, user_id, timestamp, from_status, to_status) VALUES (
+          $taskId, $userId, now(), $initialStatus, $newStatus
+          )""".update.run
+    } else {
+      0.pure[ConnectionIO]
+    }
+
+  def getTaskById(taskId: UUID): ConnectionIO[Option[Task]] =
+    query.filter(taskId).selectOption
+
+  def unsafeGetTaskById(taskId: UUID): ConnectionIO[Task] =
+    query.filter(taskId).select
+
+  def getTaskActions(taskId: UUID): ConnectionIO[List[TaskActionStamp]] = {
+    Dao
+      .QueryBuilder[TaskActionStamp](
+        fr"select task_id, user_id, timestamp, from_status, to_status FROM task_actions",
+        fr"task_actions",
+        Nil
+      )
+      .filter(fr"task_id = $taskId")
+      .list
+  }
+
+  def getTaskWithActions(taskId: UUID): ConnectionIO[Option[Task.TaskFeature]] =
+    (for {
+      task <- OptionT(getTaskById(taskId))
+      actions <- OptionT.liftF(getTaskActions(task.id))
+    } yield { task.toGeoJSONFeature(actions) }).value
+
+  def unsafeGetTaskWithActions(taskId: UUID): ConnectionIO[Task.TaskFeature] =
+    for {
+      task <- unsafeGetTaskById(taskId)
+      actions <- getTaskActions(task.id)
+    } yield { task.toGeoJSONFeature(actions) }
+
+  def updateTask(
+      taskId: UUID,
+      updateTask: Task.TaskFeatureCreate,
+      user: User
+  ): ConnectionIO[Option[Task.TaskFeature]] =
+    (for {
+      initial <- OptionT(getTaskById(taskId))
+      _ <- OptionT.liftF(updateF(taskId, updateTask).update.run)
+      _ <- OptionT.liftF(
+        appendAction(
+          taskId,
+          initial.status,
+          updateTask.properties.status,
+          user.id
+        )
+      )
+      withActions <- OptionT(getTaskWithActions(taskId))
+    } yield withActions).value
+
+  def deleteTask(taskId: UUID): ConnectionIO[Int] =
+    query.filter(taskId).delete
+
+  def listTasks(
+      queryParams: TaskQueryParameters,
+      pageRequest: PageRequest
+  ): ConnectionIO[PaginatedGeoJsonResponse[Task.TaskFeature]] =
+    for {
+      paginatedResponse <- Dao
+        .QueryBuilder[Task](listF, joinTableF, Nil)
```

Comment:
Hmm, so, to me, it looks like `listF` is already a `cols ++ joinTableF`? So I guess you meant `cols` in here?

Reply:
No -- the `tableF` parameter to querybuilder isn't actually used in the list query, it's just used for things like building `countF`

Reply:
@jisantuc makes sense to me now.